// turbowatch/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.watch = exports.defineConfig = exports.TurboWatcher = exports.FSWatcher = exports.FileWatchingBackend = exports.ChokidarWatcher = void 0;
var ChokidarWatcher_1 = require("./backends/ChokidarWatcher");
Object.defineProperty(exports, "ChokidarWatcher", { enumerable: true, get: function () { return ChokidarWatcher_1.ChokidarWatcher; } });
var FileWatchingBackend_1 = require("./backends/FileWatchingBackend");
Object.defineProperty(exports, "FileWatchingBackend", { enumerable: true, get: function () { return FileWatchingBackend_1.FileWatchingBackend; } });
var FSWatcher_1 = require("./backends/FSWatcher");
Object.defineProperty(exports, "FSWatcher", { enumerable: true, get: function () { return FSWatcher_1.FSWatcher; } });
var TurboWatcher_1 = require("./backends/TurboWatcher");
Object.defineProperty(exports, "TurboWatcher", { enumerable: true, get: function () { return TurboWatcher_1.TurboWatcher; } });
var defineConfig_1 = require("./defineConfig");
Object.defineProperty(exports, "defineConfig", { enumerable: true, get: function () { return defineConfig_1.defineConfig; } });
var watch_1 = require("./watch");
Object.defineProperty(exports, "watch", { enumerable: true, get: function () { return watch_1.watch; } });
